ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -662,78 +728,88 @@ private void readFromFile(File fileToRead, Journal journal) throws StoreExceptio
-                  logEndOffset, key, blobValue.getOffset(), blobValue.getSize(),
-                  blobValue.isFlagSet(IndexValue.Flags.Delete_Index));
-            }
-          }
-          endOffset.set(new Offset(startOffset.getName(), maxEndOffset));
-          logger.trace("IndexSegment : {} setting end offset for index {}", indexFile.getAbsolutePath(), maxEndOffset);
-          long crc = crcStream.getValue();
-          if (crc != stream.readLong()) {
-            // reset structures
-            keySize = KEY_SIZE_INVALID_VALUE;
-            valueSize = VALUE_SIZE_INVALID_VALUE;
-            endOffset.set(startOffset);
-            index.clear();
-            bloomFilter.clear();
-            throw new StoreException("IndexSegment : " + indexFile.getAbsolutePath() + " crc check does not match",
-                StoreErrorCodes.Index_Creation_Failure);
-          }
+          persistedEntrySize = keySize + valueSize;
+          indexSizeExcludingEntries = VERSION_FIELD_LENGTH + KEY_OR_ENTRY_SIZE_FIELD_LENGTH + VALUE_SIZE_FIELD_LENGTH
+              + LOG_END_OFFSET_FIELD_LENGTH + CRC_FIELD_LENGTH;
+          break;
+        case PersistentIndex.VERSION_2:
+          persistedEntrySize = stream.readInt();
+          valueSize = stream.readInt();
+          indexSizeExcludingEntries = VERSION_FIELD_LENGTH + KEY_OR_ENTRY_SIZE_FIELD_LENGTH + VALUE_SIZE_FIELD_LENGTH
+              + LOG_END_OFFSET_FIELD_LENGTH + CRC_FIELD_LENGTH;
           break;
         default:
           throw new StoreException("IndexSegment : " + indexFile.getAbsolutePath() + " invalid version in index file ",
               StoreErrorCodes.Index_Version_Error);
       }
+      long logEndOffset = stream.readLong();
+      if (version == PersistentIndex.VERSION_0) {
+        lastModifiedTimeSec.set(indexFile.lastModified() / 1000);
+      } else if (version == PersistentIndex.VERSION_1 || version == PersistentIndex.VERSION_2) {
+        lastModifiedTimeSec.set(stream.readLong());
+        StoreKey storeKey = factory.getStoreKey(stream);
+        short resetKeyType = stream.readShort();
+        resetKey = new Pair<>(storeKey, PersistentIndex.IndexEntryType.values()[resetKeyType]);
+        indexSizeExcludingEntries +=
+            (LAST_MODIFIED_TIME_FIELD_LENGTH + resetKey.getFirst().sizeInBytes() + RESET_KEY_TYPE_FIELD_LENGTH);
+      }
+      firstKeyRelativeOffset = indexSizeExcludingEntries - CRC_FIELD_LENGTH;
+      logger.trace("IndexSegment : {} reading log end offset {} from file", indexFile.getAbsolutePath(), logEndOffset);
+      long maxEndOffset = Long.MIN_VALUE;
+      while (stream.available() > CRC_FIELD_LENGTH) {
+        StoreKey key = factory.getStoreKey(stream);
+        byte[] value = new byte[valueSize];
+        stream.read(value);
+        if (version == PersistentIndex.VERSION_2) {
+          byte[] padding = new byte[persistedEntrySize - valueSize - key.sizeInBytes()];
+          stream.read(padding);
```

Comment:
Your comment made me look at the doc and it seems that we should not be using `stream.read()` at all to begin with as it can, for various reasons, read less than asked. We should instead use `stream.readFully()`. I have made the change here for reading the value. As for skipping over padded bytes, neither `skip(n)` nor `skipBytes(n)` provide the guarantee that they will return only after reading n bytes. Therefore made a change to use an array that is created once (similar to your other suggestion) and then use `readFully()` to read off all the bytes of the padding.